### ISSUE TITLE
[Syntax highlighting] Custom scopes are usually not implemented by themes

### ISSUE THREAD

**Issue by ephread** [sequitur/ink-vscode#4]
Since Ink is a relatively niche language, most (if not all) the popular color themes for VS Code do not implement its custom scopes, resulting in a somewhat partial syntax highlighting. [TODO.md](https://github.com/sequitur/ink-vscode/blob/master/TODO.md) does acknowledge the issue:

> Include theme files with comprehensive colors specifically for Ink.

I can only think of two and half possible options to solve the issue:

1. fork all the most popular themes, add support for the custom scopes and bundle them with the extension — that's what seems to be implied in the TODO, but it would most likely be a significant undertaking;

1. (1.5) do 1., but also create PRs and praying for them to be accepted, so that users won't have to switch themes when writing in Ink;

2. rework the Ink `.tmLanguage` by renaming custom copes to most widely accepted ones, slightly sacrificing the original semantics — the `.tmLanguage` would also have to periodically merge and adapt the changes from upstream (luckily, the syntax is not likely to change much).

Since the vast majority of available themes for VS Code implement _at least_ the most common scopes found in the original [TypeScript .tmLanguage](https://github.com/Microsoft/TypeScript-TmLanguage), I'd be enclined to go with 3., which seems to be a good compromise between user-friendliness and maintainability.

What are your thoughts on this?

**Comment by sequitur** [comment]
Currently, the `.tmLanguage` file uses scopes that semantically match common general scopes, sometimes with Ink-specific detail at the bottom level of the hierarchy (like `keyword.operator.divert`). This means most themes work reasonably, but customizing them to make them work better (by differentiating `*` operators from `->` operators, for example) is easy and involves invoking semantic names.

I'm happy with this solution, but the upstream `.tmLanguage` isn't ideal about this and could be tweaked for better highlighting with standard themes. Some major things that jump out are not treating `[` and `{` brackets as operators (which they currently are, whereas they should be punctuation), renaming the scope used in `~` lines, and giving the `~` character itself its own scope.

**Comment by ephread** [comment]
I had a specific and simple example in mind, which I failed to provide to illustrate my thoughts. Even when not going into intricate constructs, there are some issues. For instance, with the following knot and variable declaration, the syntax highlighting doesn't seem quite correct.

```
=== prologue_dream ===
VAR asked_who_are_you = false
```

Below are the resulting highlights, with four different themes (stock theme in Inky, default VS Code Dark+, Solarized Dark and Material Palenight):

<img width="275" alt="theme" src="https://user-images.githubusercontent.com/722366/42222632-126f2c3e-7f19-11e8-86be-e184ebb983eb.png">

Most of the tokens are not highlighted, something I attribute to the choice of scopes. To focus on the variable declaration, the scope matching the variable name would be better expressed as `variable.other` rather than `entity.name.variable` (in this particular case, there is also another issue at play, as the assignment part doesn't seem to be matched by any scope).

Since I'm not too familiar with TextMate grammars, I originally thought that the problems I saw with different themes only stemmed from misnamed scopes (_misnamed in relation to the conventions followed by VS Code themes_) and that there was some specifics involved with VS Code. After your explanation and upon further inspection I realize that the issues are more far reaching.

Thanks for helping me understand this a bit better! I think we can close the issue, as it’s actually not something specific to `ink-vscode`. I'll open an issue later on the main ink repository, and see what can be done about it.
